File: recursion/factorial.c
#include <stdio.h>

int factorial_tail(int n, int acc) {
    if (n == 0 || n == 1) {
        return acc;
    } else {
        return factorial_tail(n - 1, n * acc);
    }
}

int factorial(int n) {
    // original recursion
    // if (n == 0 || n == 1) {
    //     return 1;
    // } else {
    //     return n * factorial(n - 1);
    // }
    return factorial_tail(n, 1);
}

int main() {
    printf("%d\n", factorial(5));
    return 0;
}

File: recursion/factorial.py
def factorial(n):
  assert n>=0 and int(n)==n, "n must be a positive integer only"
  if n in [0,1]:
    return 1
  else:
    return n*factorial(n-1)
  
print(factorial(100))

# tail recursion version
def factorial_tail(n, acc=1):
    assert n >= 0 and int(n) == n, "n must be a positive integer only"
    if n in [0, 1]:
        return acc
    else:
        return factorial_tail(n - 1, n * acc)

print(factorial_tail(5))
